### DISCUSSION TITLE
v1.5.0

### DISCUSSION THREAD

**Discussion by boonhapus** [thoughtspot/cs_tools#117] [General]
## v1.5.0 - [Lots of Organization](https://github.com/thoughtspot/cs_tools/releases/tag/v1.5.0)

Read about the whole release in [Our Documentation](https://thoughtspot.github.io/cs_tools/changelog/1.5.0/).

**🛠️ Added**
- support for ThoughtSpot Orgs in Config file
- `cs_tools self sync` - to align the underlying dependencies without requiring an update of __CS Tools__ itself
- a new tool for interacting with the ThoughtSpot Git APIs
- support pulling from `ENVIRONMENT_VARIABLES` for the CS Tools config
- CS Tools Searchable additional tables (`TS_CLUSTER` , `TS_DATA_SOURCE` , `TS_ORG`, `TS_XREF_ORG`)
- CS Tools Searchable Spot App: Worksheet Column Utilization
- CS Tools Searchable Spot App: ThoughtSpot Adoption

**🔧 Modified**
- noisiness of the analytics capture ~ now with more silence ~
- gave `cs_tools tools bulk-deleter` a confirmation gate so users don't accidentally delete large quantities of content

**🐛 Bugfix**
- __tons__ of minor fixes for __Scriptability__ and parent library `thoughtspot_tml`
- fix for #63 , #64 , #66 , #71 , #72 , #81 , #93 , #94 , #95
- bugfixes to [optional] analytics data capture

**❎ Removed**
- temporarily removed BigQuery Database Syncer
- all private APIs have been removed from CS Tools

**🫀 CS Tools Internals**
- added support for V2.0 REST APIs
- the minimum supported version of __ThoughtSpot__ is now 9.0.1
- switched to `pytest` framework for testing
- switched to `ruff` for code quality
- separated the cs_tools core from the cs_tools CLI
- bump `python` version to 3.9, adding support through the latest version
- bump `httpx >= 0.27.0`
- bump `pydantic >= 2.6.4`
- bump `sqlmodel >= 0.0.16` (`sqlalchemy >= 2.0.0`)
- bump `typer >= 0.12.0`
- bump `pendulum >= 3.0.0`

---

**Full Changelog**: https://github.com/thoughtspot/cs_tools/compare/v1.4.16...v1.5.0

<hr /><em>This discussion was created from the release <a href='https://github.com/thoughtspot/cs_tools/releases/tag/v1.5.0'>v1.5.0</a>.</em>

**Comment by bmofkn** [comment]
This is great news @boonhapus!  Thanks for fostering such an awesome tool!

I just tried `cs_tools self upgrade` and while the upgrade appeared to complete successfully, I'm now getting the following when trying to run any command:
`/Library/Application Support/cs_tools/.cs_tools/lib/python3.11/site-packages/cs_tools/cli/dependencies/thoughtspot.py", line 7, in <module>
    from typer.core import TyperOption
ModuleNotFoundError: No module named 'typer.core'`

Any ideas on how to resolve?

**Reply by boonhapus** [reply]
@bmofkn Can you try the reinstall method? I added some new code in the updater which **_should_** (operative word should 😅) help going forward.

**Windows** 
```powershell
(Invoke-WebRequest `
    -Uri https://raw.githubusercontent.com/thoughtspot/cs_tools/master/cs_tools/updater/_bootstrapper.py `
    -UseBasicParsing `
).Content | python - --reinstall
```

**Mac/Linux**
```bash
curl \
    --silent --show-error --location-trusted \
    https://raw.githubusercontent.com/thoughtspot/cs_tools/master/cs_tools/updater/_bootstrapper.py \
    | python3 - --reinstall
```

**Reply by bmofkn** [reply]
That helped!  Thanks a bunch!

**Reply by bmofkn** [reply]
@boonhapus, thanks again for all of the help!  The reinstall fixed the errors with running the cs_tools command.  I can now run the following without errors:
`cs_tools self upgrade`
`cs_tools --version`


However, when I attempt to run `cs_tools self sync` or any of the `cs_tools tools scriptability` style commands, I get errors:

`cs_tools self sync`
`<my-macos-user-folder-here>/Library/Application Support/cs_tools/.cs_tools/lib/python3.11/site-packages/cs_tools/updater/_updater.py", line 113, in run
    raise RuntimeError(f"Failed with exit code: {proc.returncode}\n\nPIP COMMAND BELOW\n{cmd}")
RuntimeError: Failed with exit code: 1`

`cs_tools tools scriptability`
`<my-macos-user-folder-here>/Library/Application Support/cs_tools/.cs_tools/lib/python3.11/site-packages/cs_tools/cli/tools/scriptability/tmlfs.py", line 112, in _create_log_path
    if not self.path.exists():
           ^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'exists'`

Any ideas regarding these new issues?

**Reply by boonhapus** [reply]
@bmofkn You shouldn't need `cs_tools self sync` yet. This is to update the underlying `thoughtspot_tml` library without me requiring to push an update for CS Tools itself. I'll look into this one though.. not sure why you're finding that.

Let me look into `_create_log_path` issue as well... I need to chat with a colleague internally to fix this one.

**Reply by bmofkn** [reply]
adding to the errors... 

As an additional test, I tried to run `cs_tools tools searchable gather --config prod_ts --syncer sqlite://database_path=csToolsUsersGroupsX.db`

It ran through the whole script :) ... until the attempt to write all of the combined data to the target syncer

`/Library/Application Support/cs_tools/.cs_tools/lib/python3.11/site-packages/cs_tools/cli/tools/searchable/app.py", line 447, in metadata
    syncer.dump(model.__tablename__, data=[model.validated_init(**row).model_dump() for row in rows])
    ^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'dump'`

I'm sure it's related and doubt you need the info... posted just in case you do :)

**Reply by boonhapus** [reply]
Ah! Thanks @bmofkn This looks like a change in `typer` upstream with command aliases. I recently renamed the `gather` command to `metadata` to better reflect the activity its performing.. but it looks like once I bumped `typer` to 0.12.0 , something's changed in the invocation logic. Can you try using it this way?

`cs_tools tools searchable metadata --config prod_ts --syncer sqlite://database_path=csToolsUsersGroupsX.db`

**Reply by bmofkn** [reply]
@boonhapus, that worked!  Thanks a bunch.

**Reply by bmofkn** [reply]
@boonhapus, again, thanks for all of the support!  Sincerely appreciate it.  

While v1.5 is being "patched", is there a way to roll back to the previous stable version (was it _1.4.15_??).  We run `cs_tools tools scriptability export --include-types xxx` for answers, liveboards, connections, etc., everyday to store in a git repo, and we are not able to do that today given the upgrade to v1.5.

**Reply by boonhapus** [reply]
@bmofkn Hmm you could activate the virtual environment (find the path with `cs_tools self info`) and explicitly install cs_tools from the last version.

(after activate the venv)
`python -m pip install https://github.com/thoughtspot/cs_tools/archive/v1.4.16.zip`

It won't be simple to downgrade. I am not entirely positive how well this will work though, as we upgraded configuration files with 1.5.0.

I plan to work on a fix tonight for your issue with scriptability though. I'm thinking the simplest changes should at least unblock the workflow.

**Reply by bmofkn** [reply]
Thanks!  We will patiently wait for the patch 🙂

**Reply by boonhapus** [reply]
@bmofkn can you try updating from 1.5.0 to 1.5.1 with `cs_tools self update` ? I've tested this locally on a couple of machines as well as against 3 internal clusters. Hoping I've fixed the issues you're seeing. 🤞

**Reply by bmofkn** [reply]
@boonhapus, I ran `cs_tools self update` and it successfully updated to 1.5.1 👍

I then ran several `cs_tools tools scriptability export --include-types xxx` commands and they all completed successfully 🙌.  It looks like we are "back to normal" 🙂

Thanks so much for providing a remedy.  Sincerely appreciate it!
